drivers/tee/optee/optee.c

```diff
@@ -747,30 +790,141 @@ static int optee_get_version(const struct device *dev, struct tee_version_info *
 	return 0;
 }
 
+static size_t get_shm_size(const struct device *dev)
+{
+	struct optee_driver_data *data = (struct optee_driver_data *)dev->data;
+	size_t shm_size = OPTEE_MSG_GET_ARG_SIZE(MAX_ARG_PARAM_COUNT);
+
+	if (data->rpc_param_count != 0) {
+		shm_size += OPTEE_MSG_GET_ARG_SIZE(data->rpc_param_count);
+	}
+
+	return shm_size;
+}
+
+static int alloc_shm_cache_entry(struct shm_cache_entry **entry)
+{
+	if (entry == NULL) {
+		return -EINVAL;
+	}
+
+	*entry = k_malloc(sizeof(**entry));
+	if (*entry == NULL) {
+		return -ENOMEM;
+	}
+
+	memset(*entry, 0, sizeof(**entry));
+
+	return 0;
+}
+
+#if defined(CONFIG_OPTEE_CLEAN_SHM_CACHE)
+static void rm_shm_if_empty_unlocked(const struct device *dev, struct shm_cache_entry *entry)
+{
+	struct optee_driver_data *data = (struct optee_driver_data *)dev->data;
+	size_t args_cnt;
+	size_t shm_size = get_shm_size(dev);
+
+	if (data->rpc_param_count != 0) {
+		args_cnt = OPTEE_MSG_NONCONTIG_PAGE_SIZE / shm_size;
+	} else {
+		args_cnt = 1;
+	}
+
+	if (tee_rm_shm(dev, entry->shm)) {
+		LOG_ERR("Unable to clean shared memory");
+	}
+
+	sys_dlist_remove(&entry->node);
+	k_free(entry);
+}
+#endif /* CONFIG_OPTEE_CLEAN_SHM_CACHE */
+
+static int get_or_alloc_shm(const struct device *dev, size_t num_params, unsigned int flags,
+			    struct shm_cache_entry **entry_r, struct optee_msg_arg **arg)
+{
+	int rc = 0;
+	struct optee_driver_data *data = (struct optee_driver_data *)dev->data;
+	size_t shm_size = get_shm_size(dev);
+	size_t args_cnt;
+	struct shm_cache_entry *entry;
+
+	if (num_params > MAX_ARG_PARAM_COUNT) {
+		return -EINVAL;
+	}
+
+	if (data->rpc_param_count) {
+		args_cnt = OPTEE_MSG_NONCONTIG_PAGE_SIZE / shm_size;
+	} else {
+		args_cnt = 1;
+	}
+
+	k_mutex_lock(&data->shm_cache.mutex, K_FOREVER);
+	SYS_DLIST_FOR_EACH_CONTAINER(&data->shm_cache.shm, entry, node) {
+		if (!entry->used) {
+			goto found;
+		}
+	}
+
+	rc = alloc_shm_cache_entry(&entry);
+	if (rc != 0) {
+		goto unlock;
+	}
+
+	rc = tee_add_shm(dev, NULL, MAX_SHM_SIZE, MAX_SHM_SIZE, flags, &entry->shm);
+	if (rc != 0) {
+		LOG_ERR("Unable to get shared memory, rc = %d", rc);
+		k_free(entry);
+		goto unlock;
+	}
+
+	sys_dlist_append(&data->shm_cache.shm, &entry->node);
+found:
+	entry->used = true;
+	*arg = entry->shm->addr;
+	memset(*arg, 0, shm_size);
+	*entry_r = entry;
+unlock:
+	k_mutex_unlock(&data->shm_cache.mutex);
+	return rc;
+}
+
+static void free_shm_entry(const struct device *dev, struct shm_cache_entry *entry)
+{
+	struct optee_driver_data *data = (struct optee_driver_data *)dev->data;
+
+	if (entry == NULL) {
+		return;
+	}
+
+	k_mutex_lock(&data->shm_cache.mutex, K_FOREVER);
+	entry->used = false;
+
+#if defined(CONFIG_OPTEE_CLEAN_SHM_CACHE)
+	rm_shm_if_empty_unlocked(dev, entry);
```

Comment:
Well, `rm_shm_if_empty_unlocked` tries to remove `entry` unconditionally, if I got this right. What is the point of such "caching" then? Is it just to save RPC allocation trip? But RPC buffers are cached via different mechanism..